dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/dataset/FileLabelsService.java

```diff
@@ -0,0 +1,139 @@
+     * from the original label the entry is marked as affected. At the end
+     * all new labels are checked against the rest in order to find and handle
+     * duplicates.
+     */
+    public List<FileLabelInfo> changeLabels(List<FileLabelInfo> inputLabels, FileLabelsChangeOptionsDTO options) {
+        List<FileLabelInfo> changedLabels = new ArrayList<>();
+        for (FileLabelInfo labelInfo : inputLabels) {
+            if (!labelInfo.isAffected()) {
+                changedLabels.add(labelInfo);
+                continue;
+            }
+            String fileLabel = labelInfo.getLabel();
+            String changed = fileLabel.replaceAll(options.getFrom(), options.getTo());
+            boolean affected = !fileLabel.equals(changed);
+            changedLabels.add(new FileLabelInfo(labelInfo.getId(), fileLabel,
+                    affected ? changed : labelInfo.getLabelAfterChange(), affected));
+        }
+        // Handle duplicated names: first extract labels that aren't going to change to a map, then for each
+        // changed label check whether it's a duplicate – if so keep generating new labels until it's not.
+        // Then add this new label into the map, and check the next label.
+        Map<String, FileLabelInfo> finalLabels = new HashMap<>(changedLabels.stream()
+                .filter(e -> !e.isAffected())
+                .collect(Collectors.toMap(FileLabelInfo::getLabel, e -> e, (prev, next) -> next)));
+        for (FileLabelInfo labelInfo : changedLabels) {
+            if (!labelInfo.isAffected()) {
+                continue;
+            }
+            String newLabel = labelInfo.getLabelAfterChange();
+            while (finalLabels.containsKey(newLabel)) {
+                newLabel = IngestUtil.generateNewFileName(newLabel);
+            }
+            finalLabels.put(newLabel, new FileLabelInfo(labelInfo.getId(), labelInfo.getLabel(), newLabel, true));
+        }
+        return finalLabels.values().stream()
+                .sorted()
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * The method receives the output from {@link FileLabelsService#changeLabels(List, FileLabelsChangeOptionsDTO)}
+     * and saves the labels marked as affected into the draft version of the
+     * given dataset. The caller of the method should ensure that the draft
+     * is properly created before the method is called, otherwise issues with
+     * database constraints will follow.
+     */
+    @Restricted(@PermissionNeeded(needs = {Permission.EditDataset}, allRequired = true))
+    public List<FileLabelInfo> updateDataset(@PermissionNeeded Dataset dataset, List<FileLabelInfo> labelToChange) {
+        DatasetVersion editVersion = dataset.getEditVersion();
+        Map<Long, FileLabelInfo> labelsToChangeById = labelToChange.stream()
+                .filter(FileLabelInfo::isAffected)
+                .collect(Collectors.toMap(FileLabelInfo::getId, l -> l, (prev, next) -> next));
```

Comment:
`(prev, next) -> next` means that basically we would allow for some elements to be skipped. I would remove the accumulator. And allow for exception to be thrown if we accidentally pass two `FileLabelInfo` with the same id